homeassistant/components/irm_kmi/coordinator.py

```diff
@@ -0,0 +1,90 @@
+"""DataUpdateCoordinator for the IRM KMI integration."""
+
+from datetime import timedelta
+import logging
+
+from irm_kmi_api import IrmKmiApiClientHa, IrmKmiApiError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_LATITUDE, ATTR_LONGITUDE, CONF_LOCATION
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import (
+    TimestampDataUpdateCoordinator,
+    UpdateFailed,
+)
+from homeassistant.util import dt as dt_util
+from homeassistant.util.dt import utcnow
+
+from .data import ProcessedCoordinatorData
+from .utils import get_config_value, preferred_language
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IrmKmiCoordinator(TimestampDataUpdateCoordinator[ProcessedCoordinatorData]):
+    """Coordinator to update data from IRM KMI."""
+
+    def __init__(
+        self, hass: HomeAssistant, entry: ConfigEntry, api_client: IrmKmiApiClientHa
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            config_entry=entry,
+            name="IRM KMI weather",
+            update_interval=timedelta(minutes=7),
+        )
+        self._api = api_client
+        self._location = get_config_value(entry, CONF_LOCATION)
+
+    async def _async_update_data(self) -> ProcessedCoordinatorData:
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables so entities can quickly look up their data.
+        :return: ProcessedCoordinatorData
+        """
+
+        self._api.expire_cache()
+        try:
+            await self._api.refresh_forecasts_coord(
+                {
+                    "lat": self._location[ATTR_LATITUDE],
+                    "long": self._location[ATTR_LONGITUDE],
+                }
+            )
+
+        except IrmKmiApiError as err:
+            if (
+                self.last_update_success_time is not None
+                and self.update_interval is not None
+                and self.last_update_success_time - utcnow()
+                < timedelta(seconds=2.5 * self.update_interval.seconds)
+            ):
+                _LOGGER.warning(
+                    "Error communicating with API for general forecast: %s. Keeping the old data",
+                    err,
+                )
+                return self.data
```

Comment:
when does this happen?

Reply:
This happens when the server timeouts or when it is unreachable for some reason.  I have seen this happen a few times in my logs.